account/seed.go

```diff
@@ -0,0 +1,206 @@
+//
+// one of the specific private key types are returned using the base "PrivateKeyInterface"
+// interface type to allow individual methods to be called.
+func PrivateKeyFromBase58Seed(seedBase58Encoded string) (*PrivateKey, error) {
+
+	// verify length
+	seed := util.FromBase58(seedBase58Encoded)
+	seedLength := len(seed)
+	if seedV1Length != seedLength && seedV2Length != seedLength {
+		return nil, fault.ErrInvalidSeedLength
+	}
+
+	// verify checksum
+	digest := sha3.Sum256(seed[:seedLength-checksumLength])
+	checksumStart := seedLength - seedChecksumLength
+	expectedChecksum := digest[:seedChecksumLength]
+	actualChecksum := seed[checksumStart:]
+	if !bytes.Equal(expectedChecksum, actualChecksum) {
+		return nil, fault.ErrChecksumMismatch
+	}
+
+	header := seed[:seedHeaderLength]
+	var ed25519Seed []byte // the ed25519 seed for generate key pair
+	var testnet bool       // denote the network is test net
+
+	switch {
+	case bytes.Equal(seedHeaderV1, header):
+		// copy the secret key from seed
+		var sk [secretKeyV1Length]byte
+		secretStart := seedHeaderLength + seedPrefixLength
+		copy(sk[:], seed[secretStart:])
+
+		prefix := seed[seedHeaderLength:secretStart]
+		// first byte of prefix is test/live indication
+		testnet = prefix[0] == 0x01
+
+		ed25519Seed = secretbox.Seal([]byte{}, authSeedIndex[:], &seedNonce, &sk)
+
+	case bytes.Equal(seedHeaderV2, header):
+		sk := seed[seedHeaderLength:checksumStart]
+
+		// verify valid secret key
+		if secretKeyV2Length != len(sk) || 0 != sk[16]&0x0f {
+			return nil, fault.ErrInvalidSeedLength
+		}
+
+		// parse network
+		mode := sk[0]&0x80 | sk[1]&0x40 | sk[2]&0x20 | sk[3]&0x10
+		testnet = mode == sk[15]&0xf0^0xf0
+
+		// add the seed 4 times to hash value
+		hash := sha3.NewShake256()
+		for i := 0; i < 4; i++ {
+			n, err := hash.Write(sk)
+			if err != nil {
+				return nil, err
+			}
+			if secretKeyV2Length != n {
+				return nil, fault.ErrCannotDecodeSeed
+			}
+		}
+
+		ed25519Seed = make([]byte, ed25519.SeedSize)
+		n, err := hash.Read(ed25519Seed)
+		if nil != err {
+			return nil, err
+		}
+		if ed25519.SeedSize != n {
+			return nil, fault.ErrCannotDecodeSeed
+		}
+
+	default:
+		return nil, fault.ErrInvalidSeedHeader
+	}
+
+	// generate key pair from encrypted secret key
+	_, priv, err := ed25519.GenerateKey(bytes.NewBuffer(ed25519Seed))
+	if nil != err {
+		return nil, err
+	}
+
+	privateKey := &PrivateKey{
+		PrivateKeyInterface: &ED25519PrivateKey{
+			Test:       testnet,
+			PrivateKey: priv,
+		},
+	}
+	return privateKey, nil
+}
+
+// NewSeedV1 - generate base58 seed v1
+func NewSeedV1(testnet bool) (string, error) {
+	// generate new seed
+	seedCore := make([]byte, secretKeyV1Length)
+	n, err := rand.Read(seedCore)
+	if nil != err {
+		return "", err
+	}
+	if secretKeyV1Length != n {
+		panic("too few random bytes")
+	}
+	net := 0x00
+	if testnet {
+		net = 0x01
+	}
+	packedSeed := append(seedHeaderV1, byte(net))
```

Comment:
I think the base58 seed that we used is just an encoded seed (base58 encoded seed), it’s not a seed.  I can encode the seed in different encode algorithm even base64 or something as long as i can decode to get the raw seed. So I think the `seed` is better than `packedSeed` and the base58 seed is `encodedSeed`

Reply:
so `seed` is the plain binary bytes representation
and `encodedSeed` is base58 string representation
OK, just needs to be applied consistently

Reply:
I pushed a change, any better?

Reply:
fixed